src/layers/mlp.jl

```diff
@@ -1,16 +1,47 @@
 """
-    mlp_block(planes, hidden_planes; dropout = 0., dense = Dense, activation = gelu)
+		mlp_block(inplanes, hidden_planes; outplanes = inplanes, dropout = 0., activation = gelu)
 
-Feedforward block used in many vision transformer-like models.
+Feedforward block used in many MLPMixer-like and vision-transformer models.
 
 # Arguments
-- `planes`: Number of dimensions in the input and output.
+- `inplanes`: Number of dimensions in the input.
 - `hidden_planes`: Number of dimensions in the intermediate layer.
+- `outplanes`: Number of dimensions in the output - by default it is the same as `inplanes`.
 - `dropout`: Dropout rate.
-- `dense`: Type of dense layer to use in the feedforward block.
 - `activation`: Activation function to use.
 """
-function mlp_block(planes, hidden_planes; dropout = 0., dense = Dense, activation = gelu)
-  Chain(dense(planes, hidden_planes, activation), Dropout(dropout),
-        dense(hidden_planes, planes), Dropout(dropout))
+function mlp_block(inplanes, hidden_planes; outplanes = inplanes, dropout = 0., activation = gelu)
+  Chain(Dense(inplanes, hidden_planes, activation), Dropout(dropout),
+        Dense(hidden_planes, outplanes), Dropout(dropout))
+end
+
+"""
+    gated_mlp(inplanes, hidden_planes; outplanes = inplanes, dropout = 0., activation = gelu, 
+							gate_layer = identity)
+
+Feedforward block based on the implementation in the paper "Pay Attention to MLPs".
+([reference](https://arxiv.org/abs/2105.08050))
+
+# Arguments
+- `inplanes`: Number of dimensions in the input.
+- `hidden_planes`: Number of dimensions in the intermediate layer.
+- `outplanes`: Number of dimensions in the output - by default it is the same as `inplanes`.
+- `dropout`: Dropout rate.
+- `activation`: Activation function to use.
+- `gate_layer`: Layer to use for the gating.
+"""
+function gated_mlp(inplanes, hidden_planes; outplanes = inplanes, dropout = 0., activation = gelu, 
+									 gate_layer = identity)
+	layers = []
+	push!(layers, Dense(inplanes, hidden_planes, activation))
+	push!(layers, Dropout(dropout))
+	if typeof(gate_layer) != typeof(identity)
+		@assert hidden_planes % 2 == 0 "`hidden_planes` must be even for gated MLP"
+		gate_layer = gate_layer(hidden_planes)
+		hidden_planes ÷= 2
+	end
```

Comment:
Same suggestion here. Make `gate_layer` a 1-arg constructor, then you can just write out the `Chain` instead of pushing and splatting.

Reply:
In that case, do I move the `@assert` and the `hidden_planes` check + mutation to the inside of the constructor as well?

Reply:
Could it just be at the start of this function? Or is there a case with `identity` where `hidden_planes` isn't even? Otherwise we can just enforce it for all cases.

Reply:
Yeah `hidden_planes` will probably be odd in some cases - it's calculated programmatically so can't really be sure about that. I'll move the checks to the start, but what about the `hidden_planes` mutation?

Reply:
Do you need to mutate? You could just pass `hidden_planes ÷ 2` to the next layer. I think you are correct that having the check inside the gate layer constructor is better than the start of the function.

Also, what about combining this with `mlp_block`? The `identity` case is the same as `mlp_block` right?

Reply:
> You could just pass `hidden_planes ÷ 2` to the next layer.

Yes but this is subject to `gate_layer` not being identity

> what about combining this with `mlp_block`

This could be done, yeah - I kept it initially separate because of the checks

Reply:
Yep this is perfect - dispatch in Julia is really a beast 😁